Add support for label-dependent L4 egress policy

ianvernon opened the pull request:
This consists of the following changes:

* pkg/policy: add label-dependent L4 for egress
    - Rename existing functions in terms of ingress and factor out egress-related parts in now-ingress-only functions to their corresponding egress-related counterparts.
cmd: add directionality to cilium bpf policy <add,delete>
    - Add capability to specify whether policy should apply to ingress or egress to the given BPF PolicyMap for an endpoint.
* pkg/maps/policymap: add Invalid TrafficDirection
* cmd: add parseTrafficString helper
    - Add a helper which converts a given string to policymap.TrafficDirection.
* pkg/k8s: enable TestParseNetworkPolicyEgressL4AllowAll
    - Enable this unit test now that L3-dependent-L4 for egress is added.
* test:
    - Add L3-dependent L4 tests to Ginkgo runtime tests. Update existing policies to handle datapath changes (specifically those that set EgressPolicy to "always" enforce (default deny)).
* bpf: add label-based egress enforcement, including label-based L3, and label-based L3-dependent L4. This does *not* add L3-dependent L7, which is still yet to be merged by @jrajahalme.
* pkg/endpoint: use correct label access call in checkEgressAccess.

What this does *not* do:
* Test CLI for cilium bpf policy <add,delete>
* Expose egress policy tracing to end-users. I think this needs to be done with an API update with a verdict for ingress and egress accordingly. See #3394 .

```release-note
Add support label-dependent L4 egress policy
```

Signed-off by: Ian Vernon <ian@cilium.io>

Fixes: #3099

Commit:
pkg/policy: add label-dependent L4 for egress

Rename existing functions in terms of ingress and factor out egress-related
parts in now-ingress-only functions to their corresponding egress-related
counterparts.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/policy: remove unit test which disallows label-dependent l4

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
cmd: add directionality to `cilium bpf policy <add,delete>`

Add capability to specify whether policy should apply to ingress or egress to
the given BPF PolicyMap for an endpoint.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/maps/policymap: add Invalid TrafficDirection

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
cmd: add parseTrafficString helper

Add a helper which converts a given string to policymap.TrafficDirection.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
cmd: move updatePolicyKey to helpers.go

This function is used both by `cilium bpf policy add` and `cilium bpf policy
delete`,so move it to the file where helpers that are common to more than
one command are located.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/k8s: enable TestParseNetworkPolicyEgressL4AllowAll

Enable this unit test now that L3-dependent-L4 for egress is added.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/endpoint: use correct label access call in checkEgressAccess

Should be AllowsEgressLabelAccess, not AllowsIngressLabelAccess.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/runtime: enable test which was disabled due to lack of L3-dependent L4 egress enforcement

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/runtime: remove useless CIDR tests

CIDR policy does not apply to IPs within the cluster, which this test
was trying to test. So, these tests were not truly testing CIDR policy.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/runtime/manifests: change policy to be L3-dependent-L4 for egress

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/runtime: add egress L3-only test

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/runtime: add egress L4 with L3 label wildcard test

Misc. comment fixes as well.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/runtime: add L3-dependent L7 policy and commented-out tests

Tests are commented out because egress L3-dependent-L7 still needs to be added to Cilium.

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer committed:
bpf: Reduce IPv6 CIDR egress prefixes in build

The number being used before was way too big, and was causing common
build environments to exceed the verifier limits.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
bpf: Force inlining of handle_ipv6

Fixes an issue where clang 3.8 was generating "call 0" instructions.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
bpf: Apply L4 rules to external IPs

Previously, this check was erroneously applied to all traffic, which, in
the case that there were no L4 policies, would result in allowing all
traffic. Only apply this (and drop) in the case where the traffic is
egressing outside the cluster. If L4 would drop the traffic, drop it;
otherwise, defer to the CIDR check later in the function.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
bpf: Fix proxy_port lookup on egress

Signed-off-by: Joe Stringer <joe@covalent.io>

Commit:
test/runtime: fix up CIDR tests to reflect datapath changes

Need to add egress rules because we go into "always" PolicyEnforcement mode in the
CIDR tests.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/runtime/manifests: update conntrack test policies to account for label-based egress

Now that we do label-based lookups for egress in datapath, need to explicitly
add label-based egress policies because the agent's policy enforcement mode is
set to 'always' (default deny for egress) for these tests.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
api: add allowed-egress-identities to EndpointPolicy

Add list of allowed identities for egress communication.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/endpoint: populate AllowedEgressIdentities in GetPolicyModel

Signed-off by: Ian Vernon <ian@cilium.io>

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -868,7 +836,15 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+					{"matchLabels":{"%s":""}}
+				]
+			}]
+		}]`, httpd1Label, httpd2Label, ipv4OtherNet, "id.app3", httpd1Label)
```
- Could you change httpd1Label to use only one time, and used `%[1]s` instead of `%s`
- Could you set "id.app3" as a const, want to avoid the hardcoding in the code. Should be something like this: 

```
script = fmt.Sprintf(`
		[{
			"endpointSelector": {"matchLabels":{"%[1]s":""}},
			"ingress": [{
				"fromEndpoints": [
					{"matchLabels":{"%[2]s":""}}
				]
			}, {
				"fromCIDR": [
					"%[3]s"
				]
			}]
		},
		{
			"endpointSelector": {"matchLabels":{"%[4]s":""}},
			"egress": [{
				"toEndpoints": [
					{"matchLabels":{"%[1]s":""}}
				]
			}]
		}]`, httpd1Label, httpd2Label, ipv4OtherNet, "id.app3")

```

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -839,7 +799,15 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+					{"matchLabels":{"%s":""}}
+				]
+			}]
+		}]`, httpd1Label, httpd2Label, ipv4Prefix, ipv6Prefix, "id.app3", httpd1Label)
```
- Could you use "%[1]s" instead repeating httpd1Label
- Could you make "id.app3" a const to avoid hardcoding.

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -802,7 +754,15 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+					{"matchLabels":{"%s":""}}
+				]
+			}]
+		}]`, httpd1Label, httpd2Label, httpd2Label, httpd1Label)
```
- Could we use `%[1]s` instead of duplicating httpd1Label and httpd2Label?

joestringer committed:
bpf: Fix non-CT egress policy address lookup

When conntrack is disabled, the destination address in the ct_tuple is
not flipped, so egress IP->ID lookup was using the wrong address to
lookup the destination identity. Fix it up.

Signed-off-by: Joe Stringer <joe@covalent.io>

Commit:
daemon: add TODO for GH-3394

Add TODO message for egress policy tracing.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/endpoint: add GitHub issue number for egress ConnTrack work

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/policy: remove outdated TODO message

Unit tests are added for ResolveL4EgressPolicy, so remove TODO message to add unit tests.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
cilium/cmd: refactor parseTrafficString

Use a switch statement instead of if-else chain.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/endpoint: cleanup formatting of log message

Move fields in log message to be on each line for easier readability.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/k8s: remove remove duplicate port check for egress rules

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
cmd: add documentation and TODO for updatePolicyKey

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/runtime: send output of cilium-agent to logger

The kvstore test stops the cilium service, and starts the process directly. Thus, no logs are
sent to syslog; edit the command which runs the cilium-agent to send output to syslog
for debugging.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/runtime: reduce redundant arguments in policy formatting

Use string formatting index references to populate policies within test.

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer commented on test/runtime/Policies.go:
```diff
@@ -532,67 +539,6 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
 		connectivityTest(allRequests, helpers.App1, helpers.Httpd1, true)
 		connectivityTest(allRequests, helpers.App2, helpers.Httpd1, true)
-
-		By("Ingress CIDR")
```
I agree that these CIDR tests were not testing the actual CIDR policies.

If there's not already an issue for reviewing the CIDR policy coverage, we should file one.

joestringer commented on test/runtime/Policies.go:
```diff
@@ -606,6 +552,7 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
 			connectivityTest(httpRequestsPublic, app, helpers.Httpd2, true)
 		}
 		connectivityTest(allRequests, helpers.App3, helpers.Httpd1, false)
+		connectivityTest(pingRequests, helpers.App1, helpers.Httpd3, false)
```
How about success from App1->Httpd3 via curl?

joestringer reviewed (approved):
Reviewed the top commits. Minor comments.
